benchmarks/src/main/kotlin/bfschannel/BfsChannelBenchmark.kt

```diff
@@ -0,0 +1,362 @@
+        val graphNodes = randomConnectedGraph(nodes, edges)
+        writeGeneratedGraphToGrFile(graphFileName, graphNodes)
+        println("Generated $graphFileName")
+    }
+
+    override val graphFileName: String
+        get() = "out/$name.gr"
+
+    override fun parseGraphFile(graphFileName: String): List<Node> = parseGrFile(graphFileName, false)
+}
+
+class DownloadingGraphCreator(name: String, private val type: GraphType, private val url: String) : GraphCreator(name) {
+    override val graphFileName: String
+        get() = "out/$name.${type.ext()}.${if (type.gzipped()) "gz" else ""}"
+
+    override fun generateOrDownloadGraphFile(graphFileName: String) {
+        println("Downloading $graphFileName from $url")
+        val input = Channels.newChannel(URL(url).openStream())
+        val output = FileOutputStream(graphFileName)
+        output.channel.transferFrom(input, 0, Long.MAX_VALUE)
+        println("Downloaded $graphFileName")
+    }
+
+    override fun parseGraphFile(graphFileName: String): List<Node> {
+        return when (type) {
+            GraphType.GR_GZ -> parseGrFile(graphFileName, type.gzipped())
+            GraphType.TXT_GZ -> parseTxtFile(graphFileName, type.gzipped())
+        }
+    }
+}
+
+enum class GraphType {
+    GR_GZ {
+        override fun ext(): String = "gr"
+        override fun gzipped(): Boolean = true
+    },
+    TXT_GZ {
+        override fun ext(): String = "txt"
+        override fun gzipped(): Boolean = true
+    };
+
+    abstract fun ext(): String
+    abstract fun gzipped(): Boolean
+}
+
+fun randomConnectedGraph(nodes: Int, edges: Int): List<Node> {
+    require(edges >= nodes - 1)
+    // always use the same seed
+    val random = Random(0)
+    val nodesList = List(nodes) { Node(it + 1) }
+    // generate a random connected graph with `nodes-1` edges
+    createGraphStructure(nodesList, random)
+    // add `edges - nodes + 1` random edges
+    addEdges(edges - nodes + 1, nodes, nodesList, random)
+    return nodesList
+}
+
+private fun createGraphStructure(nodesList: List<Node>, random: Random) {
+    val nodesToConnect = ArrayList(nodesList)
+    var currentNode = nodesToConnect.removeAt(random.nextInt(nodesToConnect.size))
+    val visitedNodes = mutableSetOf(currentNode)
+    // create a structure for the graph, connect all the nodes
+    while (nodesToConnect.isNotEmpty()) {
+        val neighbor = nodesToConnect.removeAt(random.nextInt(nodesToConnect.size))
+        if (visitedNodes.add(neighbor)) {
+            currentNode.addNeighbour(neighbor)
+            neighbor.addNeighbour(currentNode)
+        }
+        currentNode = neighbor
+    }
+}
+
+private fun addEdges(edgesToAdd: Int, nodes: Int, nodesList: List<Node>, random: Random) {
+    repeat(edgesToAdd) {
+        while (true) {
+            val first = nodesList[random.nextInt(nodes)]
+            val second = nodesList[random.nextInt(nodes)]
+            if (first.id == second.id) continue
```

Comment:
It is better to skip `nodeList[first]` rather than trying again. You can re-write the code in the following way:
```
val second = nodesList[random.nextInt(nodes - 1)].run { if (it == first) nodes - 1 else it } 
```